File: gemm.py
# created by Vrindavan Sanap
# all rights reserved

import time
import numpy as np


N = 1024
arr = []
if __name__ == "__main__":
    def matMul(N):
        # N*N matrix
        A = np.random.randn(N, N).astype(np.float32)
        # N*N matrix
        B = np.random.randn(N, N).astype(np.float32)

        # flops=N^2 output cells with 2N compute each
        flops = N*N*2*N

        st = time.monotonic()
        C = A@B
        et = time.monotonic()

        s = (et-st)
        # print(f"{flops/1e9:.2f}:GFLOPS")
        # print(f"{s:.2f}:time")
        # print(f"{(flops/s)*1e-9:.2f} GFLOP/s")
        try:
            print(f"{(flops/s)*1e-9:.2f} GFLOP/s", f"{s:.5f}secs {i} ")
        except:
            pass
        arr.append(s)
    for i in range(20):
        matMul(N)

File: gemm.c
// Created by Vrindavan Sanap
//  All rights reserved

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <immintrin.h>
#include <assert.h>
#define N 1024

float A[N][N];
float B[N][N];
float C[N][N];

__m256 *An = (__m256 *)A;
__m256 *Bn = (__m256 *)B;
__m256 *Cn = (__m256 *)C;

double nanos(void)
{
    struct timespec ts;

    if (timespec_get(&ts, TIME_UTC) != TIME_UTC)
    {
        fputs("timespec_get failed!", stderr);
        return 0;
    }
    return 1000000000.0 * ts.tv_sec + ts.tv_nsec;
}

#define BLOCK 4
int main()
{
    printf("Running \n");
    for (int itr = 0; itr < 10; itr++)
    {
        double st = nanos();
        for (int by = 0; by < N; by += BLOCK)
        {
            for (int bx = 0; bx < N; bx += BLOCK)
            {
                for (int y = by; y < by + BLOCK; y++)
                {
                    for (int x = bx; x < bx + BLOCK; x++)
                    {

                        float acc = 0;
                        for (int k = 0; k < N; k++)
                        {
                            acc += A[y][k] * B[x][k];
                        }
                        C[y][x] = acc;
                    }
                }
            }
        }
        double flops = N * N * 2.0 * N;
        double et = nanos();
        double s = ((et - st) * 1e-9);
        printf(" %f GFLOP/s \n", (flops / s) * 1e-9);
    }
    return 0;
}
